### PR TITLE
LF: Structure Engine Errors.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR is part of the effort to improve engine Errors (See #9974). 
Here we classify all the errors the engine can report in 4 categories:

- Errors happening during package loading
- Error happening during command/transaction preprocessing
- Error happening during interpretation
- Error happening during transaction validation

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala [hunk 1]
```diff
@@ -4,73 +4,142 @@
...
+  final case class Package(subError: Package.SubError) extends Error
+
+  object Package {
+
+    sealed abstract class SubError extends Error.SubError {
+      final def toError: Package = Package(this)
+      override def detailMsg: String = msg
+    }
+
+    // TODO: get ride of Generic
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala:37]:
```suggestion
    // TODO: get rid of Generic
```
sounds more french with an `e` but I think the English spelling doesn’t have one :)

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala [hunk 2]
```diff
@@ -4,73 +4,142 @@
...
+  final case class Interpretation(subError: Interpretation.SubError) extends Error
+
+  object Interpretation {
+
+    sealed abstract class SubError extends Error.SubError {
+      final def toError: Interpretation = Interpretation(this)
+      override def detailMsg: String = msg
+    }
+
+    // TODO: get ride of Generic
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala:90]:
```suggestion
    // TODO: get rid of Generic
```

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala [hunk 3]
```diff
@@ -4,73 +4,142 @@
...
+      final def toError: Validation = Validation(this)
+      override def detailMsg: String = msg
+    }
 
-final case class ValidationError(override val msg: String)
-    extends RuntimeException(s"ValidationError: $msg", null, true, false)
-    with Error {
-  override def detailMsg: String = msg
-}
+    // TODO: get ride of Generic
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala:129]:
```suggestion
    // TODO: get rid of Generic
```

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala [hunk 4]
```diff
@@ -4,73 +4,142 @@
...
+          case _ => None
+        }
+    }
   }
 
-  def apply(description: String, details: String) = new Error {
-    val msg = description
-    val detailMsg = details
+  // Error happening during interpretation
+  final case class Interpretation(subError: Interpretation.SubError) extends Error
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala:81]: I don’t understand the weird `SubError` stuff. How is this better than (simplified)
```
sealed abstract class SError
sealed abstract class InterpretationError extends SError
case class ContractNotFound(cid: ContractId) extends InterpretationError
```

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala [hunk 5]
```diff
@@ -4,73 +4,142 @@
...
-}
+    final case class ReplayMismatch(
+        mismatch: transaction.ReplayMismatch[transaction.NodeId, Value.ContractId]
+    ) extends SubError {
+      override def msg: String = mismatch.msg
 
-final case class AuthorizationError(override val msg: String) extends Error {
-  override def detailMsg: String = msg
-}
+      override def detailMsg: String = ""
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala:140]: do we really want the detail to be empty? I thought the idea is that it is either identical to `msg` or contains more info.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala [hunk 6]
```diff
@@ -4,73 +4,142 @@
...
+  final case class Preprocessing(subError: Preprocessing.SubError) extends Error
+
+  object Preprocessing {
+
+    sealed abstract class SubError extends Error.SubError {
+      final def toError: Preprocessing = Preprocessing(this)
+      override def detailMsg: String = msg
+    }
+
+    // TODO: get ride of Generic
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala:60]:
```suggestion
    // TODO: get rid of Generic
```

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala [hunk 7]
```diff
@@ -4,73 +4,142 @@
...
+
+    /** See com.daml.lf.transaction.Transaction.DuplicateContractKey
+      * for more information.
+      */
+    final case class DuplicateContractKey(key: GlobalKey) extends SubError {
+      override def msg = s"Duplicate contract key $key"
+    }
+
+    final case class Authorization(override val msg: String) extends SubError {
+      override def detailMsg: String = msg
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala:114]:
```suggestion
```
already part of `SubError`

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Result.scala [hunk 8]
```diff
@@ -143,7 +146,7 @@ object Result {
     ResultNeedPackage(
       packageId,
       {
-        case None => ResultError(Error(s"Couldn't find package $packageId"))
+        case None => ResultError(Error.Interpretation.Generic(s"Couldn't find package $packageId"))
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Result.scala:149]: this is really a fatal error right? we should never hit this afaict. (perfectly fine to not touch in this PR)

**Review by cocreature — APPROVED** [review]
lgmt but I don’t get the `SubError` part

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala:140]: fixed.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Result.scala (hunk 8 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Result.scala:149]: Was planned for another  PR.

**Comment by oliverse-da** [comment]
fyi @daravep

**Pull request merged**